src/MongoDB/Query.c

```diff
@@ -50,64 +50,244 @@ PHONGO_API zend_class_entry *php_phongo_query_ce;
+	 * are not converting strings to 64-bit integers for 32-bit platforms. */
+
+#define PHONGO_QUERY_OPT_INT64(opt, zarr, key) \
+	if (zarr && php_array_existsc(zarr, key)) { \
+		if (!BSON_APPEND_INT64(intern->opts, (opt), php_array_fetchc_long(zarr, key))) { \
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Error appending \"%s\" option", (opt)); \
+			return false; \
+		} \
+	}
+
+#define PHONGO_QUERY_OPT_STRING(opt, zarr, key) \
+	if (zarr && php_array_existsc(zarr, key)) { \
+		if (!php_phongo_query_opts_append_string(intern->opts, (opt), (zarr), (key) TSRMLS_CC)) { \
+			return false; \
+		} \
+	}
+
+	PHONGO_QUERY_OPT_BOOL("allowPartialResults", options, "allowPartialResults");
+	PHONGO_QUERY_OPT_BOOL("awaitData", options, "awaitData");
+	PHONGO_QUERY_OPT_INT64("batchSize", options, "batchSize");
+	PHONGO_QUERY_OPT_DOCUMENT("collation", options, "collation");
+	PHONGO_QUERY_OPT_STRING("comment", options, "comment") else PHONGO_QUERY_OPT_STRING("comment", modifiers, "$comment");
+	PHONGO_QUERY_OPT_BOOL("exhaust", options, "exhaust");
+	PHONGO_QUERY_OPT_DOCUMENT("max", options, "max") else PHONGO_QUERY_OPT_DOCUMENT("max", modifiers, "$max");
+	PHONGO_QUERY_OPT_INT64("maxScan", options, "maxScan") else PHONGO_QUERY_OPT_INT64("maxScan", modifiers, "$maxScan");
+	PHONGO_QUERY_OPT_INT64("maxTimeMS", options, "maxTimeMS") else PHONGO_QUERY_OPT_INT64("maxTimeMS", modifiers, "$maxTimeMS");
+	PHONGO_QUERY_OPT_DOCUMENT("min", options, "min") else PHONGO_QUERY_OPT_DOCUMENT("min", modifiers, "$min");
+	PHONGO_QUERY_OPT_BOOL("noCursorTimeout", options, "noCursorTimeout");
+	PHONGO_QUERY_OPT_BOOL("oplogReplay", options, "oplogReplay");
+	PHONGO_QUERY_OPT_DOCUMENT("projection", options, "projection");
+	PHONGO_QUERY_OPT_BOOL("returnKey", options, "returnKey") else PHONGO_QUERY_OPT_BOOL("returnKey", modifiers, "$returnKey");
+	PHONGO_QUERY_OPT_BOOL("showRecordId", options, "showRecordId") else PHONGO_QUERY_OPT_BOOL("showRecordId", modifiers, "$showDiskLoc");
+	PHONGO_QUERY_OPT_INT64("skip", options, "skip");
+	PHONGO_QUERY_OPT_DOCUMENT("sort", options, "sort") else PHONGO_QUERY_OPT_DOCUMENT("sort", modifiers, "$orderby");
+	PHONGO_QUERY_OPT_BOOL("snapshot", options, "snapshot") else PHONGO_QUERY_OPT_BOOL("snapshot", modifiers, "$snapshot");
+	PHONGO_QUERY_OPT_BOOL("tailable", options, "tailable");
+
+	/* The "$explain" modifier should be converted to an "explain" option, which
+	 * libmongoc will later convert back to a modifier for the OP_QUERY code
+	 * path. This modifier will be ignored for the find command code path. */
+	PHONGO_QUERY_OPT_BOOL("explain", modifiers, "$explain");
+
+	/* mongoc_collection_find_with_opts() requires a non-negative limit. For
+	 * backwards compatibility, a negative limit should be set as a positive
+	 * value and default singleBatch to true. */
+	if (php_array_existsc(options, "limit") && php_array_fetchc_long(options, "limit") < 0) {
```

Comment:
To make this function smaller, I would probable recommend to split the special cases out into a function on their own.